### PR TITLE
Expose OAuth2 errors, avoid redirect loop.

### LINKED ISSUE #1775: expose errors in oauth2 authorization code grant
Right now when the oauth2 provider (ie Bitbucket) fails to grant an authorization code an `error` query parameter is returned, as defined in the oauth2 specification:
https://tools.ietf.org/html/rfc6749#section-4.1.2.1

Drone currently looks for the `code` and re-directs, without checking for the `error` query parameter. In error scenarios (misconfiguration for example) this can result in redirect loops, and by not catching and exposing the error it can be difficult to debug.

We should therefore start catching and returning the error. This is the block of code:
https://github.com/drone/drone/blob/master/remote/bitbucket/bitbucket.go#L46:L50

Which could be adjusted to something like this:

``` diff
package bitbucket

import (
+   "errors"
    "fmt"
    "net/http"
    "net/url"
```

``` diff
    code := r.FormValue("code")
+   if err := r.FormValue("error"); err != "" {
+       return nil, errors.New(err)
+   }
    if len(code) == 0 {
        http.Redirect(w, r, config.AuthCodeURL("drone"), http.StatusSeeOther)
        return nil, nil
    }
```

Note that this should be adjusted in both the GitHub and Bitbucket implementations.

### PR DESCRIPTION
**Pull request by AlekSi**
Closes #1775.

### CODE REVIEW

File: remote/bitbucket/bitbucket.go [hunk 1]
```diff
@@ -39,13 +40,27 @@ func New(client, secret string) remote.Remote {
...
+	if err := req.FormValue("error"); err != "" {
+		description := req.FormValue("error_description")
+		if description != "" {
+			err += " " + description
+		}
+		uri := req.FormValue("error_uri")
+		if uri != "" {
+			err += " " + uri
+		}
+		return nil, errors.New(err)
```
Comment (tboerger) [inline remote/bitbucket/bitbucket.go]: Just use fmt.Error

File: remote/bitbucket/bitbucket.go [hunk 2]
```diff
@@ -39,13 +40,27 @@ func New(client, secret string) remote.Remote {
 
 // Login authenticates an account with Bitbucket using the oauth2 protocol. The
 // Bitbucket account details are returned when the user is successfully authenticated.
-func (c *config) Login(w http.ResponseWriter, r *http.Request) (*model.User, error) {
-	redirect := httputil.GetURL(r)
+func (c *config) Login(w http.ResponseWriter, req *http.Request) (*model.User, error) {
```
Comment (tboerger) [inline remote/bitbucket/bitbucket.go]: This change is not really related to the fix.

Comment (AlekSi) [inline remote/bitbucket/bitbucket.go]: That's for style unification: other remotes use `req` for request and `r` for `*model.Repo`. Should I revert this change?

**Comment by AlekSi** [comment]
> Generally I would prefer a simple error message and only log the details.

But that would introduce a dependency on logger. Is it ok?

**Comment by bradrydzewski** [comment]
> Generally I would prefer a simple error message and only log the details.
> But that would introduce a dependency on logger. Is it ok?

agreed that we don't want to log in the `remote.github` package. In this case it would probably be better to create a `remote.AuthError` that holds the error and description. The logging could then be pushed to the handler.


```
if aerr, ok := err.(remote.AuthError); ok {
    log.Errorf(aerr.Desc)
}
```

but overall this is a great fix, and I really appreciate you taking the time to put this patch together. I'll get this merged this afternoon.

**Comment by tboerger** [comment]
Looks good so far, but which error gets exposed to the user now?

**Comment by bradrydzewski** [comment]
@tboerger right now drone does not capture an error at all, resulting in a redirect loop. See https://github.com/drone/drone/issues/1775

**Comment by AlekSi** [comment]
Yeah, in my case that was due to `http` vs `https` in URL (absent `X-Forwarded-Proto` in nginx config). Wasn't easy to debug.

**Comment by tboerger** [comment]
@bradrydzewski but with that pr the redirect loop should be avoided?

**Comment by AlekSi** [comment]
Bump.

**Comment by bradrydzewski** [comment]
wow, sorry, I thought I had merged this already. Thanks again!

**Pull request merged**